kratos/mpi/utilities/gather_modelpart_utility.cpp

```diff
@@ -182,6 +183,256 @@ template void GatherModelPartUtility::GatherOnMaster(const Variable<array_1d<dou
+    )
+{
+    /* First make all partitions aware of which communications are needed (with the current information we only know the entities of we want to bring in current partition) */
+
+    // Entity name
+    std::string entity_name;
+    if constexpr (std::is_same<TObjectType, Node>::value) {
+        entity_name = "Node";
+    } else if constexpr (std::is_same<TObjectType, Element>::value) {
+        entity_name = "Element";
+    } else if constexpr (std::is_same<TObjectType, Condition>::value) {
+        entity_name = "Condition";
+    } else {
+        KRATOS_ERROR << "Entity type not supported" << std::endl;
+    }
+
+    // Retrieve MPI data
+    const auto& r_data_communicator = rModelPart.GetCommunicator().GetDataCommunicator();
+    const int rank = r_data_communicator.Rank();
+    const int world_size = r_data_communicator.Size();
+
+    // First counting how many entities transfer for partition
+    int tag_send = 1;
+    std::vector<int> other_partition_indices;
+    other_partition_indices.reserve(world_size - 1);
+    for (int i_rank = 0; i_rank < world_size; ++i_rank) {
+        if (i_rank != rank) other_partition_indices.push_back(i_rank);
+    }
+    std::map<int, std::vector<std::size_t>> send_entities;
+    for (int i_rank = 0; i_rank < world_size; ++i_rank) {
+        if (i_rank == rank) {
+            for (auto index : other_partition_indices) {
+                std::vector<std::size_t> send_vector;
+                r_data_communicator.Recv(send_vector, index, tag_send);
+                // Just adding in case not empty
+                if (send_vector.size() > 0) {
+                    send_entities[index] = send_vector;
+                }
+            }
+        } else {
+            auto it_find = rEntitiesToBring.find(i_rank);
+            // Sending in case defined
+            if (it_find != rEntitiesToBring.end()) {
+                r_data_communicator.Send(it_find->second, i_rank, tag_send);
+            } else { // Sending empty vector in case not defined
+                std::vector<std::size_t> empty_vector;
+                r_data_communicator.Send(empty_vector, i_rank, tag_send);
+            }
+        }
+    }
+
+    // Depending of the echo level, print info
+    if (EchoLevel > 0) {
+        std::stringstream buffer;
+        buffer << "\nRank " << rank << " has to bring " << entity_name << "s from other partitions:" << std::endl;
+        for (auto bring : rEntitiesToBring) {
+            buffer << "\tFrom rank " << bring.first << ": " << bring.second.size() << " " << entity_name << "s\n\t" << bring.second << std::endl;
+        }
+        buffer << "\nRank " << rank << " has to send " << entity_name << "s from other partitions:" << std::endl;
+        for (auto send : send_entities) {
+            buffer << "\tTo rank " << send.first << ": " << send.second.size() << " " << entity_name << "s\n\t" << send.second << std::endl;
+        }
+        buffer << std::endl;
+        std::cout << buffer.str();
```

Comment:
Use of KRATOS_INFO? Would be nice to have the ability to turn them off with logger status as well.

Reply:
Done